### ISSUE TITLE
RecursionError when sum is used as right hand side of modulo

### ISSUE THREAD

**Issue by DRMacIver** [sympy/sympy#14563]
The following code fails with a RecursionError:

```python
from sympy.abc import x
from sympy import Sum

print(1 % Sum(1, (x, 0, 1)))
```

Error and some stack trace:

```
    retval = cfunc(*args, **kwargs)
  File venv/lib/python3.6/site-packages/sympy/core/numbers.py", line 1276, in __eq__
    other = _sympify(other)
  File venv/lib/python3.6/site-packages/sympy/core/sympify.py", line 401, in _sympify
    return sympify(a, strict=True)
  File venv/lib/python3.6/site-packages/sympy/core/sympify.py", line 266, in sympify
    if a in sympy_classes:
  File venv/lib/python3.6/site-packages/sympy/core/numbers.py", line 1367, in __hash__
    return super(Float, self).__hash__()
  File venv/lib/python3.6/site-packages/sympy/core/numbers.py", line 739, in __hash__
    return super(Number, self).__hash__()
RecursionError: maximum recursion depth exceeded while calling a Python object
```

Those line numbers are from current master (38b233c605489d6e64738b2da99248f677f4a29f). I've also verified this on `1.1.1`.

**Comment by asmeurer** [comment]
Dang, this actually gives `Fatal Python error: Cannot recover from stack overflow.` for me (in Python 3.5).

**Comment by asmeurer** [comment]
For me the stack trace ends in

```
  File "./sympy/core/expr.py", line 3320 in round
  File "./sympy/core/expr.py", line 3320 in round
  File "./sympy/core/expr.py", line 3320 in round
  File "./sympy/core/expr.py", line 3320 in round
  File "./sympy/core/expr.py", line 3320 in round
  File "./sympy/core/expr.py", line 3320 in round
  ...
/Users/aaronmeurer/anaconda3/bin/pythonw: line 3: 30197 Abort trap: 6           /Users/aaronmeurer/anaconda/python.app/Contents/MacOS/python "$@"
```

**Comment by sidhantnagpal** [comment]
Yes, I also get stack overflow.
`RecursionError: maximum recursion depth exceeded while calling a Python object`

Though `print(1 % Sum(1, (x, 0, 1)).doit())` produces the correct result.

**Comment by jksuom** [comment]
`Mod.eval` does not check that its arguments are numbers and Python seems to become confused when trying to convert the non-number `r = p/q = 1/Sum(1, (x, 0, 1))` to an integer (https://github.com/sympy/sympy/blob/master/sympy/core/mod.py#L63).

**Comment by oscarbenjamin** [comment]
I get:
```julia
In [4]: from sympy.abc import x 
   ...: from sympy import Sum 
   ...:  
   ...: print(1 % Sum(1, (x, 0, 1)))                                                                                              
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
~/current/sympy/sympy/sympy/core/evalf.py in evalf_sum(expr, prec, options)
   1191         if b != S.Infinity or a != int(a):
-> 1192             raise NotImplementedError
   1193         # Use fast hypergeometric summation if possible

NotImplementedError: 

During handling of the above exception, another exception occurred:

RecursionError                            Traceback (most recent call last)
<ipython-input-4-73543ccd5379> in <module>
      2 from sympy import Sum
      3 
----> 4 print(1 % Sum(1, (x, 0, 1)))

~/current/sympy/sympy/sympy/core/decorators.py in __sympifyit_wrapper(a, b)
     89                 if not hasattr(b, '_op_priority'):
     90                     b = sympify(b, strict=True)
---> 91                 return func(a, b)
     92             except SympifyError:
     93                 return retval

~/current/sympy/sympy/sympy/core/decorators.py in binary_op_wrapper(self, other)
    130                     else:
    131                         return f(self)
--> 132             return func(self, other)
    133         return binary_op_wrapper
    134     return priority_decorator

~/current/sympy/sympy/sympy/core/expr.py in __rmod__(self, other)
    204     @call_highest_priority('__mod__')
    205     def __rmod__(self, other):
--> 206         return Mod(other, self)
    207 
    208     @_sympifyit('other', NotImplemented)

~/current/sympy/sympy/sympy/core/cache.py in wrapper(*args, **kwargs)
     92             def wrapper(*args, **kwargs):
     93                 try:
---> 94                     retval = cfunc(*args, **kwargs)
     95                 except TypeError:
     96                     retval = func(*args, **kwargs)

~/current/sympy/sympy/sympy/core/function.py in __new__(cls, *args, **options)
    447 
    448         evaluate = options.get('evaluate', global_evaluate[0])
--> 449         result = super(Function, cls).__new__(cls, *args, **options)
    450         if evaluate and isinstance(result, cls) and result.args:
    451             pr2 = min(cls._should_evalf(a) for a in result.args)

~/current/sympy/sympy/sympy/core/cache.py in wrapper(*args, **kwargs)
     92             def wrapper(*args, **kwargs):
     93                 try:
---> 94                     retval = cfunc(*args, **kwargs)
     95                 except TypeError:
     96                     retval = func(*args, **kwargs)

~/current/sympy/sympy/sympy/core/function.py in __new__(cls, *args, **options)
    268 
    269         if evaluate:
--> 270             evaluated = cls.eval(*args)
    271             if evaluated is not None:
    272                 return evaluated

~/current/sympy/sympy/sympy/core/mod.py in eval(cls, p, q)
     90                     break
     91 
---> 92         rv = doit(p, q)
     93         if rv is not None:
     94             return rv

~/current/sympy/sympy/sympy/core/mod.py in doit(p, q)
     62             r = p/q
     63             try:
---> 64                 d = int(r)
     65             except TypeError:
     66                 pass

~/current/sympy/sympy/sympy/core/expr.py in __int__(self)
    247         if not self.is_number:
    248             raise TypeError("can't convert symbols to int")
--> 249         r = self.round(2)
    250         if not r.is_Number:
    251             raise TypeError("can't convert complex to int")

~/current/sympy/sympy/sympy/core/expr.py in round(self, p)
   3375         if not x.is_real:
   3376             i, r = x.as_real_imag()
-> 3377             return i.round(p) + S.ImaginaryUnit*r.round(p)
   3378         if not x:
   3379             return x

~/current/sympy/sympy/sympy/core/expr.py in round(self, p)
   3375         if not x.is_real:
   3376             i, r = x.as_real_imag()
-> 3377             return i.round(p) + S.ImaginaryUnit*r.round(p)
   3378         if not x:
   3379             return x

~/current/sympy/sympy/sympy/core/expr.py in round(self, p)
   3375         if not x.is_real:
   3376             i, r = x.as_real_imag()
-> 3377             return i.round(p) + S.ImaginaryUnit*r.round(p)
   3378         if not x:
   3379             return x

~/current/sympy/sympy/sympy/core/expr.py in round(self, p)
   3375         if not x.is_real:
   3376             i, r = x.as_real_imag()
-> 3377             return i.round(p) + S.ImaginaryUnit*r.round(p)
   3378         if not x:
   3379             return x

~/current/sympy/sympy/sympy/core/expr.py in round(self, p)
   3375         if not x.is_real:
   3376             i, r = x.as_real_imag()
-> 3377             return i.round(p) + S.ImaginaryUnit*r.round(p)
   3378         if not x:
   3379             return x

... last 5 frames repeated, from the frame below ...

~/current/sympy/sympy/sympy/core/expr.py in round(self, p)
   3375         if not x.is_real:
   3376             i, r = x.as_real_imag()
-> 3377             return i.round(p) + S.ImaginaryUnit*r.round(p)
   3378         if not x:
   3379             return x

RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by 0sidharth** [comment]
This has been fixed on master - 
```py
In [1]: 1 % Sum(1, (x, 0, 1))
Out[1]: 1
```
It was fixed in, 6888f72042d35d65f5ef14bee389d93c3ba9636f where assumptions was added to Sums. It looks like the bug was fixed in the commit, so perhaps this could be closed after adding a regression test.
